### PR TITLE
Usage, Key Features and Built-in validators

### PR DESCRIPTION
**Pull request by xRyuki-i**
What's in this PR:
- Removed the Table of Content feature since the [TOC] wasn't working
- A filled in the Usage Section and Key Features
- Described and Validated Built-in validators

### CODE REVIEW

File: readme.md [hunk 1]
```diff
@@ -77,14 +315,190 @@ ___
...
+| ----------- | ----------- | ----------- |
+| *number* | number | Maximum value that entered data needs to be smaller than. |
+
+```typescript
+max(number, customError?)
+```
+
+```typescript
+// Invalid value
+expect(max(12)(15)).toEqual(`Value 15 should be smaller than 12`);
```
Comment (ludbek) [inline readme.md:424]: Can we use validators along with respective decoders?

File: readme.md [hunk 2]
```diff
@@ -26,49 +26,287 @@ or
...
 ___
 
 ## Kitchen Sink
-
+```typescript
+import v, { Typeof } from 'validatex'
+
+const userSchema = v.array(
+  v.object({
+    name: v.string(),
```
Comment (ludbek) [inline readme.md]: Can we have some validators too?
e.g. v.string(v.minlength(1))

Comment (xRyuki-i) [reply readme.md]: I'll do that.

File: readme.md (hunk 1 above)
Comment (xRyuki-i) [reply readme.md:424]: Like using min and max validator after number decoder?

Comment (ludbek) [reply readme.md:424]: Yep. Reason being, validators will always be used with decoders.

File: readme.md [hunk 3]
```diff
@@ -26,65 +21,589 @@ or
...
+  return /[a-zA-Z][\da-zA-Z]+/.test(val) ? undefined : 'Must be alpha numeric.';
+}
+
+const username = v.string(alphaNumeric);
+expect(username('user1')).toEqual('user1');
+// value validation
+expect(() => username('123')).toThrow('Must be alpha numeric.');
+```
+
 ##### Parse
```
Comment (ludbek) [inline readme.md]: `parse` is not a validator. We can remove it from `validators` section.

File: readme.md [hunk 4]
```diff
@@ -26,65 +21,589 @@ or
...
+```typescript
+const customErrMsg = "This is a custom error message";
+
+const username = v.string(customErrMsg);
+
+expect(username.bind(username, 1)).toThrow(customErrorMsg);
+```
 
 #### Number
+The **number** decoder takes an unknown value, validates if its a number and returns an error or a number depending on the outcome of the validation.
```
Comment (ludbek) [inline readme.md]: How about 

```
The **number** decoder takes an unknown value, validates it and returns an error or a number depending on the outcome of the validation.
```

File: readme.md [hunk 5]
```diff
@@ -26,65 +21,589 @@ or
...
+```typescript
+(options?: DecoderOption<number>) => (val: unknown, context?: Context | undefined) => number
```
Comment (ludbek) [inline readme.md:375]: We need to document `DecoderOption` and `Context` too.

File: readme.md [hunk 6]
```diff
@@ -26,65 +21,589 @@ or
...
+
+#### Object
+The **object** decoder parses any value, validates if its value is of type Schema and returns an error or a Schema value depending on the outcome of the validation.
+
+```typescript
+type Schema = Record<string, Decoder | Switch<any>>;
+```
+**Signature of object decoder**
+```typescript
+(schema: T, option?: ObjectOption<Schema>) => (rawData: any, context?: Context) => Schema
```
Comment (ludbek) [inline readme.md]: We can just say `ObjectOption`.

File: readme.md [hunk 7]
```diff
@@ -26,65 +21,589 @@ or
...
+  unknownField1: 'unknown field 1',
+  unknownField2: 'unknown field 2',
+};
+expect(userSchema.bind(userSchema, data)).toThrow(`{"unknownField1":"Unknown field.","unknownField2":"Unknown field."}`)
+```
+
+#### Partial
+
+```typescript
+(schema: T, option?: Validator<Schema>) => (val: any, context?: Context) => Schema
```
Comment (ludbek) [inline readme.md]: I think we should replace `Validator<Schema>` with `PartialOption`

File: readme.md [hunk 8]
```diff
@@ -26,65 +21,589 @@ or
...
+```
+
+#### Array
+The **array** decoder parses any value, validates if its value is of type Arrary and returns an error or an Array value depending on the outcome of the validation.
+```typescript
+type Decoder = (value: any, context?: Context) => any;
+```
+**Signature of object decoder**
+```typescript
+(schema: T, option?: Validator<ReturnType<T>[]>) => (val: any, context?: Context) => Array
```
Comment (ludbek) [inline readme.md]: Lets replace `Validator<Returntype<T>>` with `ArrayOptions`

File: readme.md (hunk 5 above)
Comment (xRyuki-i) [reply readme.md:375]: Context is already documented on line 117, the code portion. Do I describe it? If i do then what should the description say.

**Review by ludbek — APPROVED** [review]

**Pull request merged**